Improve sig_handler function

TomasKorbar opened the pull request:
Wait until all currently running callbacks are finished and then
break the base loop.
Do not exit from sig_handler.
This approach has the advantage that we should be out of all
critical sections and memcached elegantly exits with return
from main.
Another advantage is that it does not matter which thread will
be assigned with executing sig handler because exit will be
always done from the root thread and thus we can avoid further
desynchronizations.

Resolves #669

dormando commented:
thanks!

What exit code does that end up using? normally it's EXIT_SUCCESS.

So there's the stop_main_loop flag that the graceful system uses. If the event_base_loop() for main_base exits normally it'll try to do a graceful shutdown, which should only be done deliberately.

Maybe stop_main_loop needs to be an enum, with 0 being NOT_STOP, 1 being GRACE_STOP, 2 being EXIT_NORMALLY? Then we can gracefully stop all the threads but not issue the restart_mmap_close() unless the correct signal was handled.

unless I'm misunderstanding how the base loop will exit here.

TomasKorbar commented:
> thanks!
> 
> What exit code does that end up using? previously it was EXIT_SUCCESS.
> 
In 1.5.16 it was still EXIT_SUCCESS
> So there's the stop_main_loop flag that the graceful system uses. If the event_base_loop() for main_base exits normally it'll try to do a graceful shutdown, which should only be done deliberately.
> 
Ohh. I was doing this change originaly for 1.5.16 and did not notice that there is the new stop_main_loop flag. Although that seems to me like active waiting :/ . Maybe the approach with 
event_base_loopexit is better because there is no polling.
I'll explain how the event_base_loopexit works.
When you call event_base_loopexit with NULL as the time argument than the loop will stop processing new events and let all currently running callbacks finish (which prevents segfaults in ssl library). After that event_base_loop which has been called with 0 as flags returns 0, as loop exited without any error. That allowed memcached 1.5.16 to exit as nothing special happened and return from main EXIT_SUCCESS.
> Maybe stop_main_loop needs to be an enum, with 0 being NOT_STOP, 1 being GRACE_STOP, 2 being EXIT_NORMALLY? Then we can gracefully stop all the threads but not issue the restart_mmap_close() unless the correct signal was handled.
> 
> unless I'm misunderstanding how the base loop will exit here.
If i'm right than if you would choose this way than stop_main_loop would lose its purpose. I can edit this pullrequest accordingly if you think too that this is the better solution:)
Appologies for my overlook.

TomasKorbar commented:
PR updated.

dormando commented:
Not quite there: I did that flag on purpose. Memcached doesn't use libevent's internal locks, instead event bases are pinned to particular threads. Calling event_base_loopexit() from the wrong thread can be fatal as it adds an event into the list chain to fire later and exit the loop. Even if you're in the same thread I'm fairly sure it's not safe to call that from a signal handler. It might be more safe than what you had before but still not safe.

The main_base is only used for new connections, not any worker threads. Using EVLOOP_ONCE ensures that all pending events/callbacks/etc are handled before it bounces out of the loop. The simple check in my outer loop is nearly identical to the `while (!done) {` inside event_base_loop. main_base also handles the clock timer so it's guaranteed to run every second. It's not really polling anything.

The "safe" equivalent without using EVLOOP_ONCE would to run a recurring timer and call the loopexit() from within there after checking a variable, which now looks like my flag check but with a lot more code :)

dormando commented:
Your change also still doesn't deal with the original issue; SIGUSR1 is used to tell memcached to gracefully shutdown for restart mode. With this you'd end up in a graceful stop under the wrong signal.

Sorry it's pretty complicated :) Make sure you develop against the `next` branch, since I can't exactly go back and patch 1.5.16.

TomasKorbar commented:
Hi @dormando ,
Thanks for your patience with this:)
Now when you have explained the subject more, i see that my solution was wrong. I updated PR and changed base to the next branch.

dormando commented:
Looks better, thanks!

dormando commented:
Out in 1.6.6

jefyt commented:
Confirmed that this PR also fixed the parallel TLS test issue (#666) and valgrind TLS test issue (https://github.com/memcached/memcached/issues/668#issuecomment-625413582). Test results:

1. [Parallel TLS Test](https://github.com/jefyt/memcached-windows/runs/670217949?check_suite_focus=true)
2. [Valgrind TLS Test](https://github.com/jefyt/memcached-windows/runs/670251117?check_suite_focus=true)
